File: src/aletharsis/identify.py
"""Content signatures take precedence over extension hints."""

import codecs
import io
from pathlib import Path
import zipfile

TEXT_TYPES = {
    ".txt": "text/plain", ".md": "text/markdown", ".rst": "text/x-rst",
    ".csv": "text/csv", ".json": "application/json", ".xml": "application/xml",
    ".html": "text/html", ".htm": "text/html",
}


def encoding_for(data: bytes) -> tuple[str, str | None]:
    # Check UTF-32 first: its LE BOM begins with the UTF-16 LE BOM.
    for marker, encoding in ((codecs.BOM_UTF32_LE, "utf-32-le"),
                             (codecs.BOM_UTF32_BE, "utf-32-be"),
                             (codecs.BOM_UTF8, "utf-8"),
                             (codecs.BOM_UTF16_LE, "utf-16-le"),
                             (codecs.BOM_UTF16_BE, "utf-16-be")):
        if data.startswith(marker):
            return encoding, marker.hex()
    return "utf-8", None


def identify(data: bytes, path: Path) -> tuple[str, str, str]:
    if data[:1024].lstrip().startswith(b"%PDF-"):
        return "pdf", "application/pdf", "PDF signature"
    if data.startswith((b"PK\x03\x04", b"PK\x05\x06", b"PK\x07\x08")):
        try:
            with zipfile.ZipFile(io.BytesIO(data)) as archive:
                names = set(archive.namelist())
            if {"[Content_Types].xml", "word/document.xml"} <= names:
                return "docx", "application/vnd.openxmlformats-officedocument.wordprocessingml.document", "OOXML archive members"
        except (zipfile.BadZipFile, ValueError):
            pass
        return "zip", "application/zip", "ZIP signature"
    for signature in (b"\x89PNG", b"\xff\xd8\xff", b"GIF87a", b"GIF89a", b"\x7fELF", b"\x1f\x8b", b"\xd0\xcf\x11\xe0"):
        if data.startswith(signature):
            return "binary", "application/octet-stream", "binary signature"
    encoding, bom = encoding_for(data)
    try:
        source = data.decode(encoding)
    except UnicodeDecodeError:
        if path.suffix.lower() in TEXT_TYPES:
            return "text", TEXT_TYPES[path.suffix.lower()], "text extension; decoding requires validation"
        return "binary", "application/octet-stream", "not valid supported Unicode text"
    if not bom and source and sum(ord(c) < 32 and c not in "\t\n\r" for c in source) / len(source) > 0.1:
        return "binary", "application/octet-stream", "high binary-control density"
    head = source.lstrip("\ufeff \t\r\n").lower()
    if head.startswith(("<!doctype html", "<html")):
        return "text", "text/html", "HTML signature; source text inspection"
    if head.startswith("<?xml"):
        return "text", "application/xml", "XML declaration; source text inspection"
    return "text", TEXT_TYPES.get(path.suffix.lower(), "text/plain"), "valid Unicode text; extension is a MIME hint"

File: internal/parsers/identify.go
package parsers

import (
	"archive/zip"
	"bytes"
	"strings"
)

var TextTypes = map[string]string{".txt": "text/plain", ".md": "text/markdown", ".rst": "text/x-rst", ".csv": "text/csv", ".json": "application/json", ".xml": "application/xml", ".html": "text/html", ".htm": "text/html"}

func Identify(data []byte, extension string) (string, string, string) {
	if bytes.HasPrefix(bytes.TrimLeft(data[:min(1024, len(data))], " \t\r\n\v\f"), []byte("%PDF-")) {
		return "pdf", "application/pdf", "PDF signature"
	}
	for _, sig := range []string{"PK\x03\x04", "PK\x05\x06", "PK\x07\x08"} {
		if bytes.HasPrefix(data, []byte(sig)) {
			if z, err := zip.NewReader(bytes.NewReader(data), int64(len(data))); err == nil {
				parts := map[string]bool{}
				for _, f := range z.File {
					parts[f.Name] = true
				}
				if parts["[Content_Types].xml"] && parts["word/document.xml"] {
					return "docx", "application/vnd.openxmlformats-officedocument.wordprocessingml.document", "OOXML archive members"
				}
			}
			return "zip", "application/zip", "ZIP signature"
		}
	}
	for _, sig := range []string{"\x89PNG", "\xff\xd8\xff", "GIF87a", "GIF89a", "\x7fELF", "\x1f\x8b", "\xd0\xcf\x11\xe0"} {
		if bytes.HasPrefix(data, []byte(sig)) {
			return "binary", "application/octet-stream", "binary signature"
		}
	}
	encoding, bom := Encoding(data)
	s, _, err := Decode(data, encoding)
	mime, known := TextTypes[extension]
	if err != nil {
		if known {
			return "text", mime, "text extension; decoding requires validation"
		}
		return "binary", "application/octet-stream", "not valid supported Unicode text"
	}
	controls, count := 0, 0
	for _, r := range s {
		count++
		if r < 32 && r != '\t' && r != '\r' && r != '\n' {
			controls++
		}
	}
	if bom == nil && count > 0 && float64(controls)/float64(count) > .1 {
		return "binary", "application/octet-stream", "high binary-control density"
	}
	head := strings.ToLower(strings.TrimLeft(s, "\ufeff \t\r\n"))
	if strings.HasPrefix(head, "<!doctype html") || strings.HasPrefix(head, "<html") {
		return "text", "text/html", "HTML signature; source text inspection"
	}
	if strings.HasPrefix(head, "<?xml") {
		return "text", "application/xml", "XML declaration; source text inspection"
	}
	if !known {
		mime = "text/plain"
	}
	return "text", mime, "valid Unicode text; extension is a MIME hint"
}
